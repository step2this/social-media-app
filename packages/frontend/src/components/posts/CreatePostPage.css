/* Create Post Page - French Automotive 1980s Clean Design */

.create-post-page {
  min-height: 100vh;
  padding: 2rem;
  background: linear-gradient(135deg, var(--tama-navy-blue) 0%, var(--tama-night-grey) 50%, var(--tama-gray-900) 100%);
  color: var(--tama-white);
}

.create-post-container {
  max-width: 800px;
  margin: 0 auto;
  position: relative;
}

/* Header */
.create-post-header {
  text-align: center;
  margin-bottom: 3rem;
}

.page-title {
  font-size: 2.5rem;
  font-weight: var(--weight-bold);
  color: var(--tama-white);
  margin-bottom: 1rem;
  font-family: var(--font-pixel);
  text-transform: uppercase;
  letter-spacing: 0.15em;
  text-align: center;
}

.page-subtitle {
  font-size: 1.1rem;
  color: var(--tama-electric-blue);
  margin: 0;
  font-family: var(--font-ui);
  text-align: center;
}

/* Form Container - Clean French Automotive Design */
.create-post-form {
  background: var(--tama-white);
  border: 3px solid var(--tama-electric-blue);
  border-radius: var(--radius-lg);
  padding: 2rem;
  box-shadow: var(--tama-depth-shadow);
  position: relative;
  margin-top: 2rem;
}

/* Form Groups */
.form-group {
  margin-bottom: 2rem;
}

.form-label {
  display: block;
  font-size: 1rem;
  font-weight: var(--weight-semibold);
  color: var(--tama-gray-800);
  margin-bottom: 0.5rem;
  text-transform: uppercase;
  letter-spacing: 0.05em;
  font-family: var(--font-pixel);
}

/* Image Upload Zone - Clean French Design */
.image-upload-zone {
  position: relative;
  border: 2px dashed var(--tama-electric-blue);
  border-radius: var(--radius-lg);
  padding: 2rem;
  text-align: center;
  background: var(--tama-gray-50);
  transition: var(--transition-normal);
  cursor: pointer;
  min-height: 200px;
  display: flex;
  align-items: center;
  justify-content: center;
}

.image-upload-zone:hover,
.drop-zone--active {
  border-color: var(--tama-racing-red);
  background: var(--tama-pearl-white);
  box-shadow: var(--tama-performance-shadow);
  transform: translateY(-1px);
}

.file-input {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
}

.upload-placeholder {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 1rem;
}

.upload-icon {
  font-size: 3rem;
  color: var(--tama-electric-blue);
  margin-bottom: 0.5rem;
}

.upload-text p {
  margin: 0.25rem 0;
  color: var(--tama-gray-700);
  font-family: var(--font-ui);
}

.upload-text p:first-child {
  font-size: 1.1rem;
  font-weight: var(--weight-semibold);
  color: var(--tama-electric-blue);
}

.upload-hint {
  font-size: 0.85rem;
  color: var(--tama-gray-500);
}

/* Image Preview */
.image-preview-container {
  position: relative;
  max-width: 400px;
  margin: 0 auto;
}

.image-preview {
  max-width: 100%;
  max-height: 300px;
  border-radius: var(--radius-md);
  border: 2px solid var(--tama-electric-blue);
  box-shadow: var(--tama-performance-shadow);
}

.remove-image-btn {
  position: absolute;
  top: -8px;
  right: -8px;
  width: 28px;
  height: 28px;
  border-radius: 50%;
  background: var(--tama-racing-red);
  border: 2px solid var(--tama-white);
  color: var(--tama-white);
  font-size: 1rem;
  font-weight: var(--weight-bold);
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  transition: var(--transition-fast);
  box-shadow: var(--tama-racing-shadow);
}

.remove-image-btn:hover {
  background: var(--tama-rally-yellow);
  color: var(--tama-gray-900);
  transform: scale(1.05);
}

/* Form Inputs - Clean French Automotive Style */
.form-input,
.form-textarea {
  width: 100%;
  padding: 0.875rem;
  border: 2px solid var(--tama-gray-300);
  border-radius: var(--radius-md);
  background: var(--tama-white);
  color: var(--tama-gray-900);
  font-size: var(--text-base);
  font-family: var(--font-ui);
  transition: var(--transition-fast);
}

.form-input:focus,
.form-textarea:focus {
  outline: none;
  border-color: var(--tama-electric-blue);
  box-shadow: 0 0 0 3px rgba(90, 172, 250, 0.2);
}

.form-input::placeholder,
.form-textarea::placeholder {
  color: var(--tama-gray-500);
  font-style: italic;
}

.form-textarea {
  resize: vertical;
  min-height: 120px;
}

/* Character Count */
.character-count {
  text-align: right;
  font-size: 0.8rem;
  color: var(--tama-gray-500);
  margin-top: 0.25rem;
  font-family: var(--font-mono);
}

/* Form Hints */
.form-hint {
  font-size: 0.8rem;
  color: var(--tama-gray-600);
  margin-top: 0.25rem;
  font-family: var(--font-ui);
}

/* Privacy Toggle - Clean Conventional Design */
.privacy-toggle {
  display: flex;
  align-items: center;
  gap: 0.75rem;
  margin-bottom: 0.5rem;
}

.privacy-checkbox {
  display: none;
}

.privacy-label {
  display: flex;
  align-items: center;
  gap: 0.5rem;
  cursor: pointer;
  user-select: none;
  font-size: var(--text-base);
  color: var(--tama-gray-800);
  font-family: var(--font-ui);
  font-weight: var(--weight-medium);
}

.toggle-switch {
  position: relative;
  width: 48px;
  height: 24px;
  background: var(--tama-gray-300);
  border: 2px solid var(--tama-gray-400);
  border-radius: 12px;
  transition: var(--transition-normal);
}

.toggle-switch::before {
  content: '';
  position: absolute;
  top: 1px;
  left: 1px;
  width: 18px;
  height: 18px;
  background: var(--tama-white);
  border-radius: 50%;
  transition: var(--transition-normal);
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.2);
}

.privacy-checkbox:checked + .privacy-label .toggle-switch {
  background: var(--tama-electric-blue);
  border-color: var(--tama-electric-blue);
}

.privacy-checkbox:checked + .privacy-label .toggle-switch::before {
  transform: translateX(22px);
  background: var(--tama-white);
}

.privacy-text {
  font-weight: var(--weight-medium);
  text-transform: none;
}

/* Submit Button - Clean French Automotive Style */
.submit-btn {
  width: 100%;
  padding: 1rem 2rem;
  font-size: var(--text-base);
  font-weight: var(--weight-semibold);
  text-transform: uppercase;
  letter-spacing: 0.05em;
  border-radius: var(--radius-md);
  transition: var(--transition-normal);
  margin-top: 1.5rem;
  font-family: var(--font-pixel);
}

.submit-btn:disabled {
  opacity: 0.6;
  cursor: not-allowed;
  transform: none;
}

.submit-btn:not(:disabled):hover {
  transform: translateY(-1px);
  box-shadow: var(--tama-racing-shadow);
}

/* Error Messages - Clean Design */
.error-message {
  color: var(--tama-racing-red);
  font-size: 0.85rem;
  margin-top: 0.5rem;
  padding: 0.5rem 0.75rem;
  background: rgba(220, 20, 60, 0.05);
  border: 1px solid rgba(220, 20, 60, 0.2);
  border-radius: var(--radius-sm);
  display: flex;
  align-items: center;
  gap: 0.5rem;
  font-family: var(--font-ui);
}

.error-message::before {
  content: '⚠️';
  flex-shrink: 0;
}

.error-message--general {
  border-color: var(--tama-racing-red);
  background: rgba(220, 20, 60, 0.05);
  margin-bottom: 1rem;
}

/* Responsive Design - Clean and Minimal */
@media (max-width: 768px) {
  .create-post-page {
    padding: 1rem;
  }

  .create-post-form {
    padding: 1.5rem;
  }

  .page-title {
    font-size: 2rem;
  }

  .image-upload-zone {
    min-height: 150px;
    padding: 1.5rem;
  }
}

@media (max-width: 480px) {
  .create-post-page {
    padding: 0.75rem;
  }

  .create-post-form {
    padding: 1rem;
  }

  .page-title {
    font-size: 1.75rem;
  }
}

/* Accessibility Enhancements */
@media (prefers-reduced-motion: reduce) {
  * {
    transition: none !important;
    animation: none !important;
  }
}

/* High contrast mode */
@media (prefers-contrast: high) {
  .form-input,
  .form-textarea {
    border-width: 3px;
    border-color: var(--tama-gray-900);
  }

  .error-message {
    border-width: 2px;
  }

  .toggle-switch {
    border-width: 3px;
  }
}