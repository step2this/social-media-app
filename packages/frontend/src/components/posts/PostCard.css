/* Post Card Component Styles */

.post-card {
  background: var(--color-surface);
  border-radius: var(--radius-lg);
  overflow: hidden;
  box-shadow: var(--shadow-sm);
  margin-bottom: var(--spacing-lg);
}

.post-card--compact {
  margin-bottom: var(--spacing-md);
}

/* Image Container */
.post-card__image-container {
  position: relative;
  width: 100%;
  aspect-ratio: 1 / 1;
  background: var(--color-border);
  overflow: hidden;
}

.post-card__image {
  width: 100%;
  height: 100%;
  object-fit: cover;
  transition: opacity 0.3s ease;
}

.post-card__image.loading {
  opacity: 0;
}

.post-card__image.loaded {
  opacity: 1;
}

.image-loading-skeleton {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: var(--color-border);
  overflow: hidden;
}

.loading-shimmer {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: linear-gradient(
    90deg,
    transparent,
    rgba(255, 255, 255, 0.1),
    transparent
  );
  animation: shimmer 2s infinite;
}

@keyframes shimmer {
  0% {
    transform: translateX(-100%);
  }
  100% {
    transform: translateX(100%);
  }
}

/* Post Info */
.post-card__info {
  padding: var(--spacing-md);
}

.post-card__header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-bottom: var(--spacing-sm);
}

.post-card__user {
  display: flex;
  align-items: center;
  gap: var(--spacing-sm);
}

.post-card__avatar {
  width: 32px;
  height: 32px;
  border-radius: 50%;
  background: var(--color-border);
}

.post-card__username {
  font-weight: 600;
  color: var(--color-text-primary);
}

.post-card__caption {
  margin: var(--spacing-sm) 0;
  color: var(--color-text-primary);
  line-height: 1.5;
}

.post-card__tags {
  display: flex;
  flex-wrap: wrap;
  gap: var(--spacing-xs);
  margin: var(--spacing-sm) 0;
}

.post-card__tag {
  display: inline-block;
  padding: var(--spacing-xs) var(--spacing-sm);
  background: var(--color-primary-light);
  color: var(--color-primary);
  border-radius: var(--radius-md);
  font-size: 0.875rem;
  font-weight: 500;
}

.post-card__timestamp {
  margin-top: var(--spacing-sm);
  font-size: 0.875rem;
  color: var(--color-text-secondary);
}

/* Actions */
.post-card__actions {
  padding: 0 var(--spacing-md) var(--spacing-md);
}

.action-buttons {
  display: flex;
  gap: var(--spacing-md);
  align-items: center;
}

.action-button-wrapper {
  display: flex;
  align-items: center;
  gap: var(--spacing-xs);
}

.action-count {
  font-size: 0.875rem;
  color: var(--color-text-secondary);
  font-weight: 500;
}

/* Comments Section */
.post-card__comments {
  padding: var(--spacing-md);
  border-top: 1px solid var(--color-border);
}

.post-card__comments-heading {
  font-size: 1rem;
  font-weight: 600;
  margin-bottom: var(--spacing-sm);
  color: var(--color-text-primary);
}

.post-card__comments-empty {
  padding: var(--spacing-md);
  text-align: center;
  color: var(--color-text-secondary);
}

/* Responsive */
@media (max-width: 768px) {
  .post-card {
    margin-bottom: var(--spacing-md);
  }

  .post-card__info {
    padding: var(--spacing-sm);
  }

  .post-card__actions {
    padding: 0 var(--spacing-sm) var(--spacing-sm);
  }
}
