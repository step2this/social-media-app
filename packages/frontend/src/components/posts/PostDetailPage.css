/* Post Detail Page - Mobile-First Vertical Layout */

.post-detail-page {
  min-height: 100vh;
  width: 100%;
  display: flex;
  justify-content: center;
  background: var(--tama-white);
}

/* Main Container - Full Width Card */
.post-detail-container {
  width: 100%;
  max-width: 100%;
  background: var(--tama-white);
  box-shadow: var(--tama-depth-shadow);
  overflow: hidden;
  color: var(--tama-gray-900);
  display: flex;
  flex-direction: column;
}

/* Close Button Overlay */
.close-button {
  position: absolute;
  top: 1rem;
  right: 1rem;
  z-index: 10;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background: rgba(0, 0, 0, 0.5);
  backdrop-filter: blur(4px);
  border: 2px solid var(--tama-white);
  color: var(--tama-white);
  cursor: pointer;
  transition: var(--transition-fast);
  display: flex;
  align-items: center;
  justify-content: center;
}

.close-button:hover {
  background: rgba(220, 20, 60, 0.8);
  transform: scale(1.1);
}

/* Main Content Layout - Vertical */
.post-content-wrapper {
  display: flex;
  flex-direction: column;
  width: 100%;
}

/* Image Section - Top Full Width */
.post-image-section {
  width: 100%;
  background: var(--tama-gray-900);
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  min-height: 400px;
  max-height: 600px;
}

.post-image-container {
  position: relative;
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.image-loading-skeleton {
  position: absolute;
  inset: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background: var(--tama-gray-200);
}

.loading-shimmer {
  width: 100%;
  height: 100%;
  background: linear-gradient(
    90deg,
    var(--tama-gray-300) 25%,
    var(--tama-gray-200) 50%,
    var(--tama-gray-300) 75%
  );
  background-size: 200% 100%;
  animation: shimmer 1.5s infinite;
}

@keyframes shimmer {
  0% {
    background-position: 200% 0;
  }
  100% {
    background-position: -200% 0;
  }
}

.post-image {
  max-width: 100%;
  max-height: 100vh;
  width: auto;
  height: auto;
  object-fit: contain;
  transition: opacity 0.3s ease;
}

.post-image.loading {
  opacity: 0;
}

.post-image.loaded {
  opacity: 1;
}

/* Content Below Image - Full Width */
.post-sidebar {
  width: 100%;
  display: flex;
  flex-direction: column;
  border-top: 2px solid var(--tama-gray-200);
}

/* Post Info Section */
.post-info-section {
  padding: 1.5rem;
  border-bottom: 2px solid var(--tama-gray-200);
}

.sidebar-user-info {
  display: flex;
  align-items: center;
  gap: 0.75rem;
  margin-bottom: 1rem;
}

.sidebar-user-avatar {
  width: 36px;
  height: 36px;
  border-radius: 50%;
  background: var(--tama-gray-300);
  border: 2px solid var(--tama-electric-blue);
}

.sidebar-user-handle {
  font-family: var(--font-pixel);
  font-weight: var(--weight-semibold);
  color: var(--tama-gray-900);
  font-size: var(--text-base);
}

.post-caption {
  font-family: var(--font-ui);
  color: var(--tama-gray-800);
  line-height: 1.6;
  margin-bottom: 1rem;
  word-wrap: break-word;
}

/* Tags */
.post-tags {
  display: flex;
  flex-wrap: wrap;
  gap: 0.5rem;
  margin-bottom: 1rem;
}

.post-tag {
  display: inline-block;
  background: var(--tama-electric-blue);
  color: var(--tama-white);
  font-family: var(--font-ui);
  font-size: var(--text-xs);
  font-weight: var(--weight-medium);
  padding: 0.25rem 0.75rem;
  border-radius: var(--radius-sm);
  border: 1px solid var(--tama-electric-blue);
  transition: var(--transition-fast);
}

.post-tag:hover {
  background: var(--tama-white);
  color: var(--tama-electric-blue);
  border-color: var(--tama-electric-blue);
}

.post-timestamp {
  font-family: var(--font-mono);
  font-size: var(--text-xs);
  color: var(--tama-gray-500);
}

/* Actions Section */
.post-actions {
  padding: 1.5rem;
  border-bottom: 2px solid var(--tama-gray-200);
}

.action-buttons {
  display: flex;
  align-items: center;
  gap: 1rem;
}

.action-button-wrapper {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.action-count {
  font-family: var(--font-ui);
  font-size: var(--text-sm);
  font-weight: var(--weight-medium);
  color: var(--tama-gray-700);
}

/* Like Button States */
.tama-btn--liked {
  color: var(--tama-racing-red);
}

.tama-btn--liked:hover {
  background-color: rgba(220, 20, 60, 0.1);
}

.tama-btn--icon:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

/* Comments Section */
.comments-section {
  flex: 1;
  padding: 1.5rem;
  overflow-y: auto;
  max-height: 400px;
}

.comments-heading {
  font-family: var(--font-pixel);
  font-weight: var(--weight-semibold);
  color: var(--tama-gray-900);
  margin-bottom: 1rem;
  font-size: var(--text-lg);
  text-transform: uppercase;
  letter-spacing: 0.05em;
}

.comments-empty {
  text-align: center;
  padding: 2rem 0;
  color: var(--tama-gray-500);
  font-family: var(--font-ui);
}

/* Add Comment Section */
.add-comment-section {
  padding: 1.5rem;
  border-top: 2px solid var(--tama-gray-200);
  background: var(--tama-pearl-white);
}

.comment-input-wrapper {
  display: flex;
  align-items: center;
  gap: 0.75rem;
}

.comment-input {
  flex: 1;
  padding: 0.75rem;
  border: 2px solid var(--tama-gray-300);
  border-radius: var(--radius-md);
  background: var(--tama-white);
  color: var(--tama-gray-900);
  font-size: var(--text-base);
  font-family: var(--font-ui);
  transition: var(--transition-fast);
}

.comment-input:focus {
  outline: none;
  border-color: var(--tama-electric-blue);
  box-shadow: 0 0 0 3px rgba(90, 172, 250, 0.2);
}

.comment-input::placeholder {
  color: var(--tama-gray-500);
  font-style: italic;
}

.comment-post-btn {
  padding: 0.75rem 1.5rem;
  background: var(--tama-electric-blue);
  color: var(--tama-white);
  border: none;
  border-radius: var(--radius-md);
  font-family: var(--font-pixel);
  font-weight: var(--weight-semibold);
  font-size: var(--text-sm);
  cursor: pointer;
  transition: var(--transition-fast);
  text-transform: uppercase;
}

.comment-post-btn:hover {
  background: var(--tama-racing-red);
  transform: translateY(-1px);
  box-shadow: var(--tama-racing-shadow);
}

/* Loading State */
.post-detail-loading {
  min-height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2rem;
  background: linear-gradient(135deg, var(--tama-navy-blue) 0%, var(--tama-night-grey) 50%, var(--tama-gray-900) 100%);
}

.loading-container {
  max-width: 1200px;
  width: 100%;
  height: 600px;
  background: var(--tama-gray-200);
  border-radius: var(--radius-lg);
  overflow: hidden;
}

.loading-shimmer-container {
  width: 100%;
  height: 100%;
  background: linear-gradient(
    90deg,
    var(--tama-gray-300) 25%,
    var(--tama-gray-200) 50%,
    var(--tama-gray-300) 75%
  );
  background-size: 200% 100%;
  animation: shimmer 1.5s infinite;
}

/* Error State */
.post-detail-error {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 2rem;
  text-align: center;
  background: linear-gradient(135deg, var(--tama-navy-blue) 0%, var(--tama-night-grey) 50%, var(--tama-gray-900) 100%);
}

.error-content {
  max-width: 600px;
}

.error-message {
  color: var(--tama-racing-red);
  font-size: var(--text-lg);
  font-family: var(--font-ui);
  font-weight: var(--weight-semibold);
  margin-bottom: 2rem;
  padding: 1rem 1.5rem;
  background: rgba(220, 20, 60, 0.05);
  border: 2px solid var(--tama-racing-red);
  border-radius: var(--radius-md);
}

/* Responsive Design */
@media (max-width: 768px) {
  .post-image-section {
    min-height: 300px;
    max-height: 500px;
  }

  .comments-section {
    max-height: 300px;
  }

  .close-button {
    width: 36px;
    height: 36px;
    top: 0.75rem;
    right: 0.75rem;
  }
}

@media (max-width: 480px) {
  .post-info-section,
  .post-actions,
  .comments-section,
  .add-comment-section {
    padding: 1rem;
  }

  .comment-input-wrapper {
    flex-direction: column;
    gap: 0.5rem;
  }

  .comment-post-btn {
    width: 100%;
  }

  .close-button {
    width: 32px;
    height: 32px;
  }
}

/* Accessibility Enhancements */
@media (prefers-reduced-motion: reduce) {
  * {
    transition: none !important;
    animation: none !important;
  }
}

/* High contrast mode */
@media (prefers-contrast: high) {
  .post-detail-container {
    border-width: 4px;
  }

  .comment-input {
    border-width: 3px;
  }

  .post-tag {
    border-width: 2px;
  }
}
