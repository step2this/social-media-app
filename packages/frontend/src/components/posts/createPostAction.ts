import { z } from 'zod';
import { CreatePostRequestSchema, type CreatePostRequest } from '@social-media-app/shared';
import { PostServiceGraphQL } from '../../services/implementations/PostService.graphql';
import { createGraphQLClient } from '../../graphql/client';
import { unwrap } from '../../graphql/types';

// Initialize post service
const postService = new PostServiceGraphQL(createGraphQLClient());

/**
 * State returned from create post action
 */
export interface CreatePostActionState {
  success: boolean;
  error: string | null;
  postId?: string;
}

/**
 * Input data for creating a post
 */
export interface CreatePostInput {
  caption: string;
  tags: string[];
  imageFile: File;
}

/**
 * Server action for creating a post
 * Handles validation, GraphQL mutation, and S3 upload
 *
 * @param prevState - Previous action state (unused but required by useActionState)
 * @param input - Post creation input (caption, tags, imageFile)
 * @returns Action state with success/error status and postId
 */
export async function createPostAction(
  prevState: CreatePostActionState,
  input: CreatePostInput
): Promise<CreatePostActionState> {
  try {
    // Validate request data
    const requestData: CreatePostRequest = {
      fileType: input.imageFile.type as 'image/jpeg' | 'image/png' | 'image/gif' | 'image/webp',
      caption: input.caption.trim(),
      tags: input.tags,
    };

    const validatedData = CreatePostRequestSchema.parse(requestData);

    // Create post via GraphQL - returns CreatePostPayload with upload URLs
    const createPayload = unwrap(await postService.createPost(validatedData));

    // Upload image to S3 using the pre-signed URL
    const uploadResponse = await fetch(createPayload.uploadUrl, {
      method: 'PUT',
      body: input.imageFile,
      headers: {
        'Content-Type': input.imageFile.type,
      },
    });

    if (!uploadResponse.ok) {
      throw new Error('Failed to upload image to S3');
    }

    // Return success with post ID
    return {
      success: true,
      error: null,
      postId: createPayload.post.id,
    };
  } catch (error) {
    console.error('Error creating post:', error);

    // Handle validation errors
    if (error instanceof z.ZodError) {
      const firstError = error.issues[0];
      return {
        success: false,
        error: `${firstError.path.join('.')}: ${firstError.message}`,
      };
    }

    // Handle other errors
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Failed to create post',
    };
  }
}
