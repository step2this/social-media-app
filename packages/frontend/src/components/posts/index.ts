export { CreatePostPage } from './CreatePostPage.js';