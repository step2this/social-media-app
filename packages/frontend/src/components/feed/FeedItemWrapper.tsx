import React from 'react';
import type { PostWithAuthor } from '@social-media-app/shared';
import { PostCard } from '../posts/PostCard';
import { useFeedItemAutoRead } from '../../hooks/useFeedItemAutoRead';

/**
 * Props for FeedItemWrapper component
 * Uses TypeScript best practices with explicit property types
 */
export interface FeedItemWrapperProps {
  /** Post data including author information */
  readonly post: PostWithAuthor;
  /** Whether to display the post in compact mode (default: true) */
  readonly compact?: boolean;
}

/**
 * Wrapper component for PostCard with auto-read functionality
 * 
 * Handles Instagram-like behavior where posts are marked as read after viewing.
 * Uses intersection observer via useFeedItemAutoRead hook to track visibility.
 * 
 * @example
 * ```tsx
 * <FeedItemWrapper post={post} compact={true} />
 * ```
 * 
 * TypeScript patterns applied from SKILL.md:
 * - Readonly modifier for immutable props
 * - Explicit interface for better error messages
 * - Optional properties with default values
 * - JSDoc comments for documentation
 */
export const FeedItemWrapper: React.FC<FeedItemWrapperProps> = ({ 
  post, 
  compact = true 
}) => {
  const elementRef = useFeedItemAutoRead(post.id);

  return (
    <div ref={elementRef as React.RefObject<HTMLDivElement>} className="feed-item-wrapper" data-testid="feed-item-wrapper">
      <PostCard post={post} compact={compact} />
    </div>
  );
};
