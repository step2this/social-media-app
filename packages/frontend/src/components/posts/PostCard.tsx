import React, { useState } from 'react';
import type { Post, FeedPostItem } from '@social-media-app/shared';
import { MaterialIcon } from '../common/MaterialIcon';
import { useLike } from '../../hooks/useLike';
import { UserLink } from '../common/UserLink';
import { FollowButton } from '../common/FollowButton';
import './PostCard.css';

export interface PostCardProps {
  post: Post | FeedPostItem;
  currentUserId?: string;
  showComments?: boolean;
  compact?: boolean;
  variant?: 'feed' | 'detail';
}

/**
 * Reusable post card component
 * Displays post with image, caption, tags, and interaction buttons
 * Consistent UI across feed and post detail pages
 * @param variant - Layout variant: 'feed' (square images, compact) or 'detail' (full images, expanded)
 */
export const PostCard: React.FC<PostCardProps> = ({
  post,
  currentUserId,
  showComments = false,
  compact = false,
  variant = 'feed'
}) => {
  const [imageLoaded, setImageLoaded] = useState(false);

  // Initialize like hook with post data
  const {
    isLiked,
    likesCount,
    isLoading: likeLoading,
    toggleLike
  } = useLike(post.id, {
    initialIsLiked: post.isLiked || false,
    initialLikesCount: post.likesCount || 0
  });

  // Determine if this is the current user's post
  const isOwnPost = currentUserId && currentUserId === post.userId;

  // Build class names based on props
  const cardClasses = [
    'post-card',
    compact && 'post-card--compact',
    variant === 'detail' && 'post-card--detail'
  ].filter(Boolean).join(' ');

  return (
    <article className={cardClasses} data-testid="post-card">
      {/* Post Image */}
      <div className="post-card__image-container">
        {!imageLoaded && (
          <div className="image-loading-skeleton">
            <div className="loading-shimmer"></div>
          </div>
        )}
        <img
          src={post.imageUrl}
          alt={post.caption || 'Post image'}
          className={`post-card__image ${imageLoaded ? 'loaded' : 'loading'}`}
          onLoad={() => setImageLoaded(true)}
          onError={() => console.error('Failed to load image')}
        />
      </div>

      {/* Post Info */}
      <div className="post-card__info">
        {/* User Info & Follow Button */}
        <div className="post-card__header">
          <div className="post-card__user">
            <div className="post-card__avatar"></div>
            <UserLink
              userId={post.userId}
              username={post.userHandle}
              className="post-card__username"
            />
          </div>
          {!isOwnPost && <FollowButton userId={post.userId} />}
        </div>

        {/* Caption */}
        {post.caption && (
          <p className="post-card__caption">{post.caption}</p>
        )}

        {/* Tags */}
        {post.tags && post.tags.length > 0 && (
          <div className="post-card__tags">
            {post.tags.map((tag, index) => (
              <span key={index} className="post-card__tag">
                #{tag}
              </span>
            ))}
          </div>
        )}

        {/* Timestamp */}
        <p className="post-card__timestamp">
          {new Date(post.createdAt).toLocaleString()}
        </p>
      </div>

      {/* Actions */}
      <div className="post-card__actions">
        <div className="action-buttons">
          {/* Like Button */}
          <div className="action-button-wrapper">
            <button
              className={`tama-btn tama-btn--icon ${isLiked ? 'tama-btn--liked' : ''}`}
              onClick={toggleLike}
              disabled={likeLoading}
              aria-label={isLiked ? 'Unlike' : 'Like'}
              data-testid="like-button"
            >
              <MaterialIcon
                name="favorite"
                variant={isLiked ? 'filled' : 'outlined'}
                size="md"
              />
            </button>
            <span className="action-count" data-testid="like-count">{likesCount}</span>
          </div>

          {/* Comment Button */}
          <div className="action-button-wrapper">
            <button
              className="tama-btn tama-btn--icon"
              aria-label="Comment"
            >
              <MaterialIcon name="chat_bubble" variant="outlined" size="md" />
            </button>
            <span className="action-count">{post.commentsCount}</span>
          </div>

          {/* Share Button */}
          <div className="action-button-wrapper">
            <button
              className="tama-btn tama-btn--icon"
              aria-label="Share"
            >
              <MaterialIcon name="share" variant="outlined" size="md" />
            </button>
          </div>
        </div>
      </div>

      {/* Comments Section (optional) */}
      {showComments && (
        <div className="post-card__comments">
          <h4 className="post-card__comments-heading">Comments</h4>
          <div className="post-card__comments-empty">
            <p>Comments coming soon...</p>
          </div>
        </div>
      )}
    </article>
  );
};
