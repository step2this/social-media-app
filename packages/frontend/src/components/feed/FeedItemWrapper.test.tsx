import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import type { PostWithAuthor } from '@social-media-app/shared';
import { FeedItemWrapper } from './FeedItemWrapper';
import { useFeedItemAutoRead } from '../../hooks/useFeedItemAutoRead';
import { PostCard } from '../posts/PostCard';

// Mock dependencies
vi.mock('../../hooks/useFeedItemAutoRead');
vi.mock('../posts/PostCard', () => ({
  PostCard: vi.fn(() => <div data-testid="post-card">Post Card</div>)
}));

describe('FeedItemWrapper', () => {
  const mockPost: PostWithAuthor = {
    id: 'post-1',
    userId: 'user-1',
    userHandle: 'testuser',
    imageUrl: 'https://example.com/image.jpg',
    caption: 'Test post content',
    likesCount: 0,
    commentsCount: 0,
    createdAt: '2024-01-01T00:00:00.000Z',
    authorId: 'user-1',
    authorHandle: 'testuser',
    authorFullName: 'Test User',
  };

  const mockRef = { current: null };

  beforeEach(() => {
    vi.mocked(useFeedItemAutoRead).mockReturnValue(mockRef as any);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders PostCard with correct props', () => {
    render(<FeedItemWrapper post={mockPost} />);
    expect(PostCard).toHaveBeenCalledWith(
      expect.objectContaining({
        post: mockPost,
        compact: true
      }),
      expect.anything()
    );
  });

  it('uses auto-read hook', () => {
    render(<FeedItemWrapper post={mockPost} />);
    expect(useFeedItemAutoRead).toHaveBeenCalledWith(mockPost.id);
  });

  it('passes post to PostCard', () => {
    render(<FeedItemWrapper post={mockPost} />);
    expect(PostCard).toHaveBeenCalledWith(
      expect.objectContaining({ post: mockPost }),
      expect.anything()
    );
  });

  it('defaults to compact=true', () => {
    render(<FeedItemWrapper post={mockPost} />);
    expect(PostCard).toHaveBeenCalledWith(
      expect.objectContaining({ compact: true }),
      expect.anything()
    );
  });

  it('respects compact prop override', () => {
    render(<FeedItemWrapper post={mockPost} compact={false} />);
    expect(PostCard).toHaveBeenCalledWith(
      expect.objectContaining({ compact: false }),
      expect.anything()
    );
  });

  it('has wrapper div with class', () => {
    const { container } = render(<FeedItemWrapper post={mockPost} />);
    const wrapper = container.querySelector('.feed-item-wrapper');
    expect(wrapper).toBeInTheDocument();
  });

  it('ref is attached to wrapper', () => {
    const { container } = render(<FeedItemWrapper post={mockPost} />);
    const wrapper = container.querySelector('.feed-item-wrapper');
    expect(wrapper).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<FeedItemWrapper post={mockPost} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders without post errors', () => {
    const { container } = render(<FeedItemWrapper post={mockPost} />);
    expect(container).toBeInTheDocument();
    expect(screen.getByTestId('post-card')).toBeInTheDocument();
  });

  it('hook called with correct post ID', () => {
    render(<FeedItemWrapper post={mockPost} />);
    expect(useFeedItemAutoRead).toHaveBeenCalledWith('post-1');
  });
});
