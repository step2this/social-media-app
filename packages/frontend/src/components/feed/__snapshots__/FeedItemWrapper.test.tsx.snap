// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FeedItemWrapper > matches snapshot 1`] = `
<div
  class="feed-item-wrapper"
  data-testid="feed-item-wrapper"
>
  <div
    data-testid="post-card"
  >
    Post Card
  </div>
</div>
`;
